### PR TITLE
PRMDR-567: Update DownloadManifestByNhsNumber lambda to handle multiple API pathways

### PR DESCRIPTION
**Pull request by abbas-khan10**

### CODE REVIEW

File: lambdas/enums/supported_document_types.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def list():
 
     @staticmethod
     def list_names() -> List[str]:
-        return [doc_type.value for doc_type in SupportedDocumentTypes.list()]
+        return [str(doc_type.value) for doc_type in SupportedDocumentTypes.list()]
```
Comment (NogaNHS) [inline lambdas/enums/supported_document_types.py:20]: What is the reason behind this change?

File: lambdas/services/document_manifest_job_service.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                    jobStatus=ZipTraceStatus.COMPLETED, url=presigned_url
+                )
+
+    def query_zip_trace(self, job_id: str) -> DocumentManifestZipTrace:
+        response = self.dynamo_service.query_with_requested_fields(
+            table_name=self.zip_trace_table,
+            index_name="JobIdIndex",
+            search_key="JobId",
+            search_condition=job_id,
+            requested_fields=list(DocumentManifestZipTrace.model_fields.keys()),
```
Comment (joefong-nhs) [inline lambdas/services/document_manifest_job_service.py:149]: For this part, it seems that the field names do not match the actual columns in dynamodb table,
`list(DocumentManifestZipTrace.model_fields.keys())` give a list of field names in snake_case,
while on the actual table, the column names are PascalCase
When we try to run with the front end, we got a 500 error due to that

File: lambdas/handlers/document_manifest_job_handler.py [hunk 3]
```diff
@@ -14,9 +17,48 @@
...
+
+    job_id = event["queryStringParameters"]["jobId"]
+
+    document_manifest_service = DocumentManifestJobService()
+    response = document_manifest_service.create_document_manifest_presigned_url(job_id)
+
+    logger.audit_splunk_info(
+        "User has downloaded Lloyd George records",
+        {"Result": "Successful download"},
+    )
```
Comment (joefong-nhs) [inline lambdas/handlers/document_manifest_job_handler.py:54]: probably better to change this log, as we can't tell whether the user is getting a completed response with url. Even for case of status = Pending it logs this line.

File: lambdas/services/document_manifest_job_service.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                is_manifest_ready = self.s3_service.file_exist_on_s3(
+                    self.zip_output_bucket, zip_trace.zip_file_location
+                )
+                if not is_manifest_ready:
+                    raise DocumentManifestJobServiceException(
+                        404, LambdaError.ManifestMissingJob
+                    )
+                presigned_url = self.s3_service.create_download_presigned_url(
+                    s3_bucket_name=self.zip_output_bucket,
+                    file_key=zip_trace.zip_file_location,
```
Comment (joefong-nhs) [inline lambdas/services/document_manifest_job_service.py:137]: same issue as above, zip_trace.zip_file_location gave the whole s3 url, not the file_key, so it generated a wrong url

File: lambdas/enums/zip_trace.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+from enum import StrEnum
+
+
+class ZipTraceStatus(StrEnum):
+    PENDING = "Pending"
```
Comment (NogaNHS) [inline lambdas/enums/zip_trace.py:5]: Add fail here

File: lambdas/services/document_manifest_job_service.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                    logger.error("Manifest documents were not found")
+                    raise DocumentManifestJobServiceException(
+                        404, LambdaError.ManifestMissingJob
+                    )
+                presigned_url = self.s3_service.create_download_presigned_url(
+                    s3_bucket_name=self.zip_output_bucket,
+                    file_key=file_key,
+                )
+                return DocumentManifestJob(
+                    jobStatus=ZipTraceStatus.COMPLETED, url=presigned_url
```
Comment (NogaNHS) [inline lambdas/services/document_manifest_job_service.py:149]: Maybe extract this to a new method

File: lambdas/handlers/document_manifest_job_handler.py [hunk 7]
```diff
@@ -14,9 +17,48 @@
 logger = LoggingService(__name__)
 
 
-@set_request_context_for_logging
 @validate_patient_id
 @validate_document_type
+def create_manifest_job(event, context):
```
Comment (NogaNHS) [inline lambdas/handlers/document_manifest_job_handler.py:22]: maybe change name to 'handle_post_request'?

File: lambdas/services/document_manifest_job_service.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            raise DocumentManifestJobServiceException(404, LambdaError.ManifestNoDocs)
+
+        if selected_document_references:
+            self.documents = self.filter_documents_by_reference(
+                selected_document_references=selected_document_references,
+            )
+
+        documents_to_download = {
+            document.file_location: document.file_name
+            for document in self.handle_duplicate_document_filenames()
```
Comment (NogaNHS) [inline lambdas/services/document_manifest_job_service.py:62]: Could we call `self.handle_duplicate_document_filenames()` outside the loop? it's a bit unclear

File: lambdas/services/document_manifest_job_service.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                [
+                    self.document_service.fetch_available_document_references_by_type(
+                        nhs_number=nhs_number,
+                        doc_type=doc_type,
+                        query_filter=UploadCompleted,
+                    )
+                    for doc_type in doc_types
+                ]
+            )
+        )
```
Comment (joefong-nhs) [inline lambdas/services/document_manifest_job_service.py:50]: Actually we've got a helper util function `flatten` for this purpose :)
`from utils.utilities import get_file_key_from_s3_url, flatten`

```suggestion
        self.documents = flatten(
            [
                self.document_service.fetch_available_document_references_by_type(
                    nhs_number=nhs_number,
                    doc_type=doc_type,
                    query_filter=UploadCompleted,
                )
                for doc_type in doc_types
            ]
        )
```

**Review by NogaNHS — APPROVED** [review]
